cmd/snap/cmd_run.go

```diff
@@ -482,26 +491,39 @@ func (x *cmdRun) straceOpts() (opts []string, raw bool, err error) {
 	return opts, raw, nil
 }
 
+func getInfoAndApp(snapName, appName string, rev snap.Revision) (*snap.Info, *snap.AppInfo, error) {
+	info, err := getSnapInfo(snapName, rev)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	app := info.Apps[appName]
+	if app == nil {
+		return nil, nil, fmt.Errorf(i18n.G("cannot find app %q in %q"), appName, snapName)
+	}
+
+	return info, app, nil
+}
+
 func (x *cmdRun) snapRunApp(snapApp string, args []string) error {
 	if x.DebugLog {
 		os.Setenv("SNAPD_DEBUG", "1")
 		logger.Debugf("enabled debug logging of early snap startup")
 	}
 	snapName, appName := snap.SplitSnapApp(snapApp)
-	info, err := getSnapInfo(snapName, snap.R(0))
+
+	snapRev := snap.R(0)
+	hint, inhibitInfo, err := runinhibit.IsLocked(snapName)
 	if err != nil {
 		return err
 	}
-
-	app := info.Apps[appName]
-	if app == nil {
-		return fmt.Errorf(i18n.G("cannot find app %q in %q"), appName, snapName)
+	if hint != runinhibit.HintNotInhibited {
+		snapRev = inhibitInfo.Previous
 	}
 
-	if !app.IsService() {
-		if err := maybeWaitWhileInhibited(snapName); err != nil {
-			return err
-		}
+	info, app, err := getInfoAndApp(snapName, appName, snapRev)
```

Comment:
notice that this code as is as the same problem as the original PR where the current symlink might not available at this point, another reason to wait before this point